sycl/include/sycl/bit_cast.hpp

```diff
@@ -24,27 +25,34 @@ namespace sycl {
 inline namespace _V1 {
 
 template <typename To, typename From>
-#if __cpp_lib_bit_cast ||                                                      \
-    (defined(__has_builtin) && __has_builtin(__builtin_bit_cast))
-constexpr
-#endif
-    std::enable_if_t<sizeof(To) == sizeof(From) &&
-                         std::is_trivially_copyable<From>::value &&
-                         std::is_trivially_copyable<To>::value,
-                     To>
-    bit_cast(const From &from) noexcept {
+constexpr std::enable_if_t<sizeof(To) == sizeof(From) &&
+                               std::is_trivially_copyable<From>::value &&
+                               std::is_trivially_copyable<To>::value,
+                           To>
+bit_cast(const From &from) noexcept {
 #if __cpp_lib_bit_cast
   return std::bit_cast<To>(from);
 #else // __cpp_lib_bit_cast
 
 #if defined(__has_builtin) && __has_builtin(__builtin_bit_cast)
   return __builtin_bit_cast(To, from);
 #else // __has_builtin(__builtin_bit_cast)
-  static_assert(std::is_trivially_default_constructible<To>::value,
+  // When building the SYCL library on Windows we may end up here.
+  // By asserting constructible and aligning the union we should be
+  // as close to free of UB as is possible without using memcpy (which is not
+  // constexpr).
+  static_assert(std::is_trivially_default_constructible_v<To>,
                 "To must be trivially default constructible");
-  To to;
-  sycl::detail::memcpy(&to, &from, sizeof(To));
-  return to;
+  static_assert(std::is_trivially_default_constructible_v<From>,
+                "From  must be trivially default constructible");
+  constexpr std::size_t max_alignment =
+      std::max(std::alignment_of_v<To>, std::alignment_of_v<From>);
+  union alignas(max_alignment) {
+    From from;
+    To to;
+  } result = {};
+  result.from = from;
+  return result.to;
```

Comment:
Are we sure the CI went through this path in any of the checks? I tried to reproduce this locally in a MWE and I'm getting this error:
```
note: read of member 'to' of union with active member 'from' is not allowed in a constant expression
   25 |   return result.to;
```

I'm testing in Linux, and is I understand this path should only be enabled in Windows, so maybe the error I'm finding does not happen in Windows, but I'd like to be sure.

Apart from the error itself, reading from the member that is not the most recently written is UB (https://en.cppreference.com/w/cpp/language/union), although, apparently, many compilers implement it. Is it safe to rely on UB?

Reply:
The CI will not go through that path with the checks.  Let me see if I can use host-compiler=cl.exe on the bit_cast tests. 

My own opinion is that the value of constexpr outweighs the UB risk. Which is low since end user code is not normally exposed to this path and it's also mitigated by the trivial default constructible requirement.

Reply:
Yes please, let's see if we can test this path at least in the tests to make sure there's no problem.

Reply:
Damnit. Good call @maarquitos14 .  That same error shows up when using host compiler.    I'll close this PR.

Have to think on this.  We definitely need some type of constexpr bit_cast or similar when building SYCL. I wonder if we have a macro for this I could guard behind?